### PR TITLE
replicaset: add `vshard bootstrap`

### PR DESCRIPTION
**Pull request by askalt**
@TarantoolBot document
Title: `tt replicaset vshard bootstrap` - bootstraps vshard in the cluster

This patch introduces new subcommand for the replicaset module.
`tt replicaset vshard` - manages vshard in the cluster.

`tt replicaset vshard bootstrap [--cartridge|--config|--custom] [flags] (<APP_NAME> | <APP_NAME:INSTANCE_NAME> | <URI>)`

* Bootstraps vshard in the replicaset.

* Cartridge: by fact, it is the clone of `cartridge-cli replicasets bootstrap-vshard`.

* Centralized config (3.0): lookups for the router in the cluster and calls `vshard.router.bootstrap()` on it.

Part of https://github.com/tarantool/tt/issues/316

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,16 @@ All notable changes to this project will be documented in this file.
...
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+- `tt replicaset expel`:  command to expel an instance from the tarantool replicaset with
+ cluster config (3.0) or cartridge orchestrator.
+- `tt replicaset vshard`: module to manage vshard in the tarantool replicaset.
+  * `tt replicaset vshard bootstrap`: command to bootstrap vshard.
+- `tt cluster replicaset expel`: command to expel an instance from the replicaset via cluster
+ config storage.
```
Comment (oleg-jukovec) [inline CHANGELOG.md:15]: Please, make a separate commit with the move of `expel` changelog entries. It will be easier to rebase/we could cherry-pick the commit in another pull request.

File: cli/cmd/replicaset.go [hunk 2]
```diff
@@ -152,6 +152,46 @@ func newExpelCmd() *cobra.Command {
 	return cmd
 }
 
+// newBootstrapVShardCmd creates a "vshard bootstrap" command.
+func newBootstrapVShardCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use: "bootstrap [--cartridge|--config] [--timeout secs] [flags] " +
```
Comment (oleg-jukovec) [inline cli/cmd/replicaset.go:158]:
```suggestion
		Use: "bootstrap [--cartridge|--config|--custom] [--timeout secs] [flags] " +
```
`addOrchestratorFlags(cmd)` will add the flag.

File: cli/replicaset/cartridge.go [hunk 3]
```diff
@@ -175,6 +180,17 @@ func (c *CartridgeInstance) Expel(ctx ExpelCtx) error {
 	return newErrExpelByInstanceNotSupported(OrchestratorCartridge)
 }
 
+// BootstrapVShard bootstraps the vshard for a single instance by the Cartridge orchestrator.
+func (c *CartridgeInstance) BootstrapVShard(ctx VShardBootstrapCtx) error {
+	err := cartridgeBootstrapVShard(c.evaler, ctx.Timeout)
+	cartridgeLogVShardBoostrapInfo(err)
```
Comment (oleg-jukovec) [inline cli/replicaset/cartridge.go:186]: Why we log something only here, but not for other subcommands?

File: test/integration/replicaset/test_vshard_app/instances.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (oleg-jukovec) [inline test/integration/replicaset/test_vshard_app/instances.yaml:1]:
```suggestion
```

File: cli/replicaset/cartridge.go (hunk 3 above)
Comment (askalt) [reply cli/replicaset/cartridge.go:186]: It's the command specific: https://github.com/tarantool/cartridge-cli/blob/master/cli/replicasets/bootstrap_vshard.go#L28-L44

Comment (oleg-jukovec) [reply cli/replicaset/cartridge.go:186]: So why not just return it as an error message?

File: test/integration/replicaset/test_vshard_app/instances.yaml (hunk 4 above)
Comment (askalt) [reply test/integration/replicaset/test_vshard_app/instances.yaml]: Removed.

File: cli/replicaset/cartridge.go (hunk 3 above)
Comment (askalt) [reply cli/replicaset/cartridge.go]: Yes, we can, did this, the error is wrapped.

File: cli/cmd/replicaset.go (hunk 2 above)
Comment (askalt) [reply cli/cmd/replicaset.go]: Added to the descrpition.

File: CHANGELOG.md (hunk 1 above)
Comment (askalt) [reply CHANGELOG.md]: Done. Thank you!

File: cli/replicaset/cconfig.go [hunk 5]
```diff
@@ -22,6 +22,11 @@ var (
 
 	//go:embed lua/cconfig/promote_election.lua
 	cconfigPromoteElectionBody string
+
+	//go:embed lua/cconfig/bootstrap_vshard_body.lua
+	cconfigBootstrapVShardBody string
+
+	cconfigGetShardingRolesBody = "return require('config'):get().sharding.roles"
```
Comment (oleg-jukovec) [inline cli/replicaset/cconfig.go:29]: I think we should add roles here for cartridge and centralized config (and print them in status).

https://github.com/tarantool/tt/blob/055990d8694a9d665d587a69df794a152ec711d8/cli/replicaset/instance.go#L7-L23

File: cli/replicaset/cconfig.go [hunk 6]
```diff
@@ -106,6 +111,17 @@ func (c *CConfigInstance) Expel(ctx ExpelCtx) error {
...
 }
 
+// BootstrapVShard bootstraps vshard for a single instance by the centralized config
+// orchestrator.
+func (c *CConfigInstance) BootstrapVShard(ctx VShardBootstrapCtx) error {
+	err := cconfigBootstrapVShard(c.evaler, ctx.Timeout)
+	if err != nil {
+		return err
+	}
+	log.Info("VShard bootstrapping is done.")
```
Comment (oleg-jukovec) [inline cli/replicaset/cconfig.go:121]: Why is the message different for Cartridge and CConfig?

**Review by oleg-jukovec — APPROVED** [review]

File: cli/replicaset/cconfig.go [hunk 7]
```diff
@@ -390,6 +406,47 @@ func (c *CConfigApplication) Demote(ctx DemoteCtx) error {
...
+	err := EvalForeach(c.runningCtx.Instances, InstanceEvalFunc(eval))
+	for _, e := range []error{err, lastErr} {
+		if e != nil {
+			return e
+		}
+	}
+	if !found {
+		return fmt.Errorf("not found any vshard router in replicaset")
+	}
+	log.Info("VShard bootstrapping is done.")
```
Comment (psergee) [inline cli/replicaset/cconfig.go:446]: I believe this message should be printed by the caller of this API.

File: test/integration/replicaset/test_replicaset_vshard.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+    rc, out = run_command_and_get_output(cmd, cwd=cartridge_app.workdir)
+    assert rc == 0
+    buf = io.StringIO(out)
+    assert "• Discovery application..." in buf.readline()
+    buf.readline()
+    # Skip init status in the output.
+    parse_status(buf)
+    assert "Bootstrapping vshard" in buf.readline()
+    assert "VShard bootstrapping is done. Check the cluster status" in buf.readline()
```
Comment (psergee) [inline test/integration/replicaset/test_replicaset_vshard.py:113]: Why do we need a message about checking the cluster status?

**Review by psergee — APPROVED** [review]
lgtm

File: cli/replicaset/cartridge.go [hunk 9]
```diff
@@ -529,6 +567,26 @@ func cartridgePromote(evaler connector.Evaler,
...
 
+// cartridgeBootstrapVShard bootstraps the vshard.
+func cartridgeBootstrapVShard(evaler connector.Evaler, timeout int) error {
+	reqOpts := connector.RequestOpts{}
+	_, err := evaler.Eval(cartridgeBootstrapVShardBody, []any{timeout}, reqOpts)
+	return err
+}
+
+// wrapCartridgeVShardBoostrapError wraps a cartridge vshard bootstrap error
+// with an addition information.
```
[the remark this answers is not in the document]
Comment (askalt) [reply cli/replicaset/cartridge.go]: Fixed.

File: cli/replicaset/cconfig.go (hunk 7 above)
Comment (askalt) [reply cli/replicaset/cconfig.go:446]: I believe in this too, I put this here because cartridge and cconfig log message are different. Let's discuss it in the comment below.

File: test/integration/replicaset/test_replicaset_vshard.py (hunk 8 above)
Comment (askalt) [reply test/integration/replicaset/test_replicaset_vshard.py:113]: @oleg-jukovec this comment relates to the [your comment about ](https://github.com/tarantool/tt/pull/820#discussion_r1574414404) messages.

It's a cartridge specific. Begins from some version it has a partial bootstrap: when you bootstrap only one of the several groups, as I understood from my investigation. And tool says to you: "it's possibly that we do the partial bootstrap, whereas you ask for the full, so, check the status".

For the details please check this https://github.com/tarantool/cartridge-cli/commit/5619178ce1cf13e2674f33e22d31eccec8cc1e1b and the mentioned issue.

File: cli/replicaset/cconfig.go (hunk 5 above)
Comment (askalt) [reply cli/replicaset/cconfig.go:29]: Let's do it in the separate task. https://github.com/tarantool/tt/issues/849

File: test/integration/replicaset/test_replicaset_vshard.py (hunk 8 above)
Comment (oleg-jukovec) [reply test/integration/replicaset/test_replicaset_vshard.py:113]: I think we need the same messages in both cases.

Comment (askalt) [reply test/integration/replicaset/test_replicaset_vshard.py]: The message is unified to the minimalistic "Done.". "Check cluster status" is removed, because it's expected behavior that cartridge can do partial bootstrap.
